src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php

```diff
@@ -109,6 +109,30 @@ public static function register($mode = 0)
             }
 
             $trace = debug_backtrace(true);
+
+            // Silence deprecation warnings about private service accessed
+            // from the service container if done so from a Test class.
+            // As of Symfony 4.1, there is a new TestContainer that allows
+            // fetching of private services within tests, so we no longer
+            // need to warn about this behavior.
+            //
+            // NOTE: the event at the top of the stack $trace (index 0) should
+            // always be the PhpUnitBridge's DeprecationErrorHandler; the
+            // second event (index 1) should be the trigger_error() event;
+            // the third event (index 2) should be the actual source of the
+            // triggered deprecation notice; and the fourth event (index 3)
+            // represents the action that called the deprecated code. In the
+            // scenario that we want to suppress, the 4th event will be an
+            // object instance of \PHPUnit\Framework\TestCase.
+            if (isset($trace[3]['object'])) {
+                $isPrivateServiceNotice = false !== strpos($msg, ' service is private, ');
+                $isNoticeForContainerGetHasUsage = 'Symfony\Component\DependencyInjection\Container' === $trace[2]['class'] && in_array($trace[2]['function'], array('get', 'has'));
+                $noticeWasTriggeredByPhpUnitTest = $trace[3]['object'] instanceof \PHPUnit\Framework\TestCase;
```

Comment:
shouldn't this test for the KernelTestCase instead (the parent TestCase does not have the client) ?

Reply:
Instanceof does it, isn't it? Not sure we need to be more strict, WDYT?

Reply:
@stof and @nicolas-grekas I didn't want to be any more strict than `\PHPUnit\Framework\TestCase` because I for one have unit tests that extend this directly instead of `KernelTestCase`.  I'm not sure if I am accessing the service container on any of those, I'd have to check, but I doubt it.  But, this might be a good point.  Is the new `TestContainer` explicitly tied to `KernelTestCase`? If so, then the deprecation warning probably shouldn't be suppressed for raw extensions of `\PHPUnit\Framework\TestCase`, as the `TestContainer` only provides access to private services via extensions of `KernelTestCase`, meaning attempts to do so via `TestCase` extensions _should_ throw errors/warnings.

If that is not the case, then I think making this comparison as general as possible by checking for extensions of `TestCase` is fine.

Reply:
Can you check this question and make the PR if needed? The reasoning lgtm.

Reply:
see: https://github.com/symfony/symfony/pull/27342